d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClient.java

```diff
@@ -13,450 +13,40 @@
    See the License for the specific language governing permissions and
    limitations under the License.
 */
-
 package com.linkedin.d2.balancer.clients;
 
+import java.util.Map;
+
+import javax.annotation.Nullable;
 
-import com.linkedin.common.callback.Callback;
-import com.linkedin.common.util.None;
 import com.linkedin.d2.balancer.LoadBalancerClient;
 import com.linkedin.d2.balancer.properties.PartitionData;
-import com.linkedin.d2.balancer.properties.PropertyKeys;
-import com.linkedin.d2.balancer.strategies.degrader.DegraderLoadBalancerStrategyConfig;
-import com.linkedin.d2.balancer.util.LoadBalancerUtil;
-import com.linkedin.data.ByteString;
-import com.linkedin.r2.RemoteInvocationException;
-import com.linkedin.r2.message.RequestContext;
-import com.linkedin.r2.message.rest.RestException;
-import com.linkedin.r2.message.rest.RestRequest;
-import com.linkedin.r2.message.rest.RestResponse;
-import com.linkedin.r2.message.stream.StreamException;
-import com.linkedin.r2.message.stream.StreamRequest;
-import com.linkedin.r2.message.stream.StreamResponse;
-import com.linkedin.r2.message.stream.entitystream.EntityStream;
-import com.linkedin.r2.message.stream.entitystream.Observer;
-import com.linkedin.r2.transport.common.bridge.client.TransportClient;
-import com.linkedin.r2.transport.common.bridge.common.TransportCallback;
-import com.linkedin.r2.transport.common.bridge.common.TransportResponse;
-import com.linkedin.util.clock.Clock;
-import com.linkedin.util.clock.SystemClock;
-import com.linkedin.util.degrader.CallCompletion;
 import com.linkedin.util.degrader.CallTracker;
-import com.linkedin.util.degrader.CallTrackerImpl;
-import com.linkedin.util.degrader.Degrader;
-import com.linkedin.util.degrader.DegraderControl;
-import com.linkedin.util.degrader.DegraderImpl;
-import com.linkedin.util.degrader.DegraderImpl.Config;
-import com.linkedin.util.degrader.ErrorType;
-
-import java.net.ConnectException;
-import java.net.URI;
-import java.nio.channels.ClosedChannelException;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
-
-import java.util.concurrent.TimeoutException;
-import java.util.regex.Pattern;
-import java.util.regex.PatternSyntaxException;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-import static com.linkedin.d2.discovery.util.LogUtil.debug;
-
-// TODO if we ever want to get rid of ties to linkedin-specific code, we'll need to move/redo call tracker/call completion/degrader
-
-public class TrackerClient implements LoadBalancerClient
+/**
+ * A client that tracks call stats and supports partitioning.
+ */
+public interface TrackerClient extends LoadBalancerClient
```

Comment:
Should we ever need to expose getCallTracker outside of tests? IMO just exposing getLatestCallStats in the interface is enough. Tests would use TrackerClientImpl directly to getCallTracker so I think the way it is now is fine.

Reply:
I do not care about the test -- we can make those functions package only. But please check all public methods in TrackerClientImpl and DegraderTrackerClient -- any one without override do not look right. The idea is that any user should use TrackerClient directly instead of the specific implementation.

Reply:
I agree that TrackerClientImpl should not be accessed through non-public methods. The situation for DegraderTrackerClient is a bit different - DegraderTrackerClient extends TrackerClientImpl, and it contains some methods that we do not invoke in the new strategy any more, but we need them in the old strategy. So you will see public methods in DegraderTrackerClient that are called by the StrategyV3, I think it's fine to keep those as public.

Reply:
Using TrackerClient interface gives us a good concept of its purpose. Otherwise having an interface in the middle does not make sense. So we should either remove TrackerClient interface completely (and use TrackerClientImpl and DegraderTrackerClient), or having a common interface and hide the implementation details behind. I'd vote for the latter.

Reply:
Will extract a DegraderTrackerClient interface.